### PR TITLE
[tests] Use ResiliencePipeline in IntegrationTests

### LINKED ISSUE #3366: [tests] Improve timeouts in EndToEnd tests
Currently, the tests use a retry+timeout mechanism on `HttpClient` like:
https://github.com/dotnet/aspire/blob/2bba2adcff99ea445971dd1ed2dfe71efaf5e959/tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs#L232-L236

.. which times out a request after 2 mins. The individual services themselves have:
https://github.com/dotnet/aspire/blob/2bba2adcff99ea445971dd1ed2dfe71efaf5e959/tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs#L18-L21

.. which will keep retrying for 60 times. But this total time can exceed the 2 mins timeout of the HttpClient on tests side, thus the client would see timeoutExceptions but not find out what happened on the service side (for example, if the container was still starting up).

- Fix this by using Resilience pipelines for services too in the `TestProject`.
- Also, collect and return all the retry failures from the service to the client to help with debugging.

cc @eerhardt

### PR DESCRIPTION
**Pull request by radical**
.. with a policy to retry with 1 sec delays, and a total timeout of
90 secs on the service side. The test(client) side has an overall http
request timeout for 120 secs which would allow the real errors from the
service to get surfaced to the client.

Fixes #3366

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3385)

### CODE REVIEW

File: tests/testproject/TestProject.IntegrationServiceA/TestUtils.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Polly;
+using Polly.Retry;
+
+namespace Aspire.TestProject;
+
+public static class TestUtils
```
Comment (radical) [inline tests/testproject/TestProject.IntegrationServiceA/TestUtils.cs:9]: better name?

File: tests/testproject/TestProject.IntegrationServiceA/SqlServer/SqlServerExtensions.cs [hunk 2]
```diff
@@ -13,14 +16,16 @@ public static void MapSqlServerApi(this WebApplication app)
 
     private static async Task<IResult> VerifySqlServerAsync(SqlConnection connection)
     {
+        StringBuilder errorMessageBuilder = new();
```
Comment (radical) [inline tests/testproject/TestProject.IntegrationServiceA/SqlServer/SqlServerExtensions.cs:19]: In a follow up PR I will add listening to logs from the services in `AppHost`, so this shouldn't be needed then.

File: tests/testproject/TestProject.IntegrationServiceA/TestProject.IntegrationServiceA.csproj [hunk 3]
```diff
@@ -32,6 +32,7 @@
     <ProjectReference Condition="'$(TestsRunningOutsideOfRepo)' != 'true'" Include="@(ComponentReferenceForTests -> '$(RepoRoot)src\Components\%(Identity)\%(Identity).csproj')" />
     <PackageReference Condition="'$(TestsRunningOutsideOfRepo)' == 'true'" Include="@(ComponentReferenceForTests)" />
     <PackageReference Include="Polly" />
```
Comment (eerhardt) [inline tests/testproject/TestProject.IntegrationServiceA/TestProject.IntegrationServiceA.csproj:34]:
```suggestion
```

We can remove this now that we are using Http.Resilience.

Actually, we shouldn't be using Http.Resilience, that is for Http calls. We should just need to change this to `Polly.Core`.

File: tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs [hunk 4]
```diff
@@ -13,15 +16,17 @@ public static void MapCosmosApi(this WebApplication app)
...
+        StringBuilder errorMessageBuilder = new();
         try
         {
-            var policy = Policy
-                .Handle<HttpRequestException>()
-                // retry 60 times with a 1 second delay between retries
-                .WaitAndRetryAsync(60, retryAttempt => TimeSpan.FromSeconds(1));
+            ResiliencePipeline pipeline = TestUtils.GetDefaultResiliencePipelineBuilder<HttpRequestException>(args =>
+            {
+                errorMessageBuilder.AppendLine(CultureInfo.InvariantCulture, $"{Environment.NewLine}Service retry #{args.AttemptNumber} due to {args.Outcome.Exception}");
```
[the remark this answers is not in the document]
Comment (radical) [reply tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs:24]: `error CA1305: The behavior of 'StringBuilder.AppendLine(ref StringBuilder.AppendInterpolatedStringHandler)' could vary based on the current user's locale settings. Replace this call in 'OracleDatabaseExtensions.VerifyOracleDatabase(MyDbContext)' with a call to 'StringBuilder.AppendLine(IFormatProvider, ref StringBuilder.AppendInterpolatedStringHandler)'. (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1305)`

[the remark this answers is not in the document]
Comment (eerhardt) [reply tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs:24]: Can we just add

https://github.com/dotnet/aspire/blob/0a28bb902c8c5271252f3f76236f5c44f371839f/src/Tools/ConfigurationSchemaGenerator/RuntimeSource/.editorconfig#L6-L7

to 

https://github.com/dotnet/aspire/blob/0a28bb902c8c5271252f3f76236f5c44f371839f/tests/.editorconfig#L1-L4

?

File: tests/testproject/TestProject.IntegrationServiceA/TestUtils.cs (hunk 1 above)
Comment (eerhardt) [reply tests/testproject/TestProject.IntegrationServiceA/TestUtils.cs:9]: Maybe ResilienceUtils?

File: tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (radical) [reply tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs]: done.

File: tests/testproject/TestProject.IntegrationServiceA/TestUtils.cs (hunk 1 above)
Comment (radical) [reply tests/testproject/TestProject.IntegrationServiceA/TestUtils.cs]: done.

File: tests/testproject/TestProject.IntegrationServiceA/ResilienceUtils.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Polly;
+using Polly.Retry;
+
+namespace Aspire.TestProject;
+
+public static class ResilienceUtils
+{
+    public static ResiliencePipelineBuilder GetDefaultResiliencePipelineBuilder<TException>(Func<OnRetryArguments<object>, ValueTask> onRetry, int overallTimeoutSecs = 90) where TException : Exception
```
Comment (eerhardt) [inline tests/testproject/TestProject.IntegrationServiceA/ResilienceUtils.cs:11]: Can you add a comment explaining the strategy here?

**Review by eerhardt — APPROVED** [review]
LGTM

**Pull request merged**